packages/xstate-react/src/useInterpret.ts

```diff
@@ -0,0 +1,117 @@
+import { useState } from 'react';
+import useIsomorphicLayoutEffect from 'use-isomorphic-layout-effect';
+import {
+  interpret,
+  EventObject,
+  StateMachine,
+  State,
+  Interpreter,
+  InterpreterOptions,
+  MachineOptions,
+  Typestate,
+  Observer
+} from 'xstate';
+import { toObserver } from 'xstate/lib/utils';
+import { MaybeLazy } from './types';
+import useConstant from './useConstant';
+import { UseMachineOptions } from './useMachine';
+import { useReactEffectActions } from './useReactEffectActions';
+
+export function useInterpret<
+  TContext,
+  TEvent extends EventObject,
+  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+>(
+  getMachine: MaybeLazy<StateMachine<TContext, any, TEvent, TTypestate>>,
+  options: Partial<InterpreterOptions> &
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>> = {},
+  observerOrListener?:
+    | Observer<State<TContext, TEvent, any, TTypestate>>
+    | ((value: State<TContext, TEvent, any, TTypestate>) => void)
+): Interpreter<TContext, any, TEvent, TTypestate> {
+  const machine = useConstant(() => {
+    return typeof getMachine === 'function' ? getMachine() : getMachine;
+  });
+
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    typeof getMachine !== 'function'
+  ) {
+    const [initialMachine] = useState(machine);
+
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+    const machineWithConfig = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    return interpret(machineWithConfig, {
+      deferEvents: true,
+      ...interpreterOptions
+    });
+  });
+
+  useIsomorphicLayoutEffect(() => {
+    let sub;
+    if (observerOrListener) {
+      sub = service.subscribe(toObserver(observerOrListener));
+    }
+
+    return () => {
+      sub?.unsubscribe();
+    };
+  }, [observerOrListener]);
```

Comment:
Given that `useMachine` uses an inline subscriber right now this effect will be refired for each commit of components using `useMachine`. I don't see any functional difference - we'll only be doing some minor extra work that shouldn't matter much and memoizing the observer also has some minor cost, so... Just something to keep in mind if it raises any problems after the release.

Reply:
Took care of this 👍 7f4caf3